### ISSUE TITLE
Issue: Custom Node with WPF Usercontrol disappears after saving 

### ISSUE THREAD

**Issue by argram30** [DynamoDS/DynamoRevit#2463]
## Dynamo version

Dynamo Core 2.0.3.8810
Dynamo Revit 2.0.3.8811

## Revit version
Autodesk® Revit® 2019
19.2.20.24
20190808_0900(x64)
2019.2.2

## Operating system

OS Name	Microsoft Windows 10 Pro for Workstations

## What did you do? 
Trying to add WPF ListBox User control to Custom Nodes, by looking at the example from here
[ButtonCustomNodeModel](https://github.com/DynamoDS/DynamoSamples/blob/master/src/SampleLibraryUI/Examples/ButtonCustomNodeModel.cs)

## What did you expect to see?

I see what I wanted to see on the Graph

![grafik](https://user-images.githubusercontent.com/15329193/71365854-871c5f00-25a0-11ea-8290-21dffda62718.png)

## What did you see instead?

But somehow only this node disappears after saving and in the console the following message appears:

```
connector 4bff7ec7-d52b-4d3e-b1d6-a99f50aed5e6 could not be created, start or end port does not exist
Active view is now Level 1
This graph has a nodeview with id:dd69af4c-f5c7-41d1-bc17-d8be1c96ad93 and name:Display RAL Color List, but does not contain a matching nodeModel
```

![grafik](https://user-images.githubusercontent.com/15329193/71366712-217da200-25a3-11ea-95d9-da3f803c267c.png)

All the other custom nodes seems to appear properly on my graph.

Here below is the Link for my code

```
using System;
using System.Collections.Generic;
using System.Collections;
using Autodesk.DesignScript.Runtime;
using DSCore;
using Dynamo.Graph.Nodes;
using ProtoCore.AST.AssociativeAST;
using RaDy.ZT.Utilities;
using RaDy.ZT.WrapperClasses;
using Newtonsoft.Json;
using System.Collections.ObjectModel;

namespace RaDy.CNM.ListBoxNodes
{
    [NodeName("Display RAL Color List")]
    [NodeDescription("Show List of RAL Colors in a List Box")]
    [IsDesignScriptCompatible]
    public class RALColorListNodeModel : NodeModel
    {
        #region Properties and Fields
        private ObservableCollection<RALColor> ralColorList;
        private readonly ObservableCollection<RALColor> defaultRALColorList = new ObservableCollection<RALColor>() { };
        
        /// <summary>
        /// Text that will appear on the button
        /// on our node.
        /// </summary>
        public ObservableCollection<RALColor> RALColorList
        {
            get { return ralColorList; }
            set
            {
                ralColorList = value;
                RaisePropertyChanged("RALColorList");
            }
        }
        #endregion

        #region constructor
        public RALColorListNodeModel()
        {
            InPorts.Add(new PortModel(PortType.Input, this, new PortData("RALCode", "RAL Codes")));
            OutPorts.Add(new PortModel(PortType.Output, this, new PortData("Color", "Dynamo Color of the RALCode")));
            RegisterAllPorts();
            
            // Listen for input port disconnection to trigger button UI update
            this.PortDisconnected += RALList_PortDisconnected;
            ArgumentLacing = LacingStrategy.Disabled;
            RALColorList = defaultRALColorList;
        }

        [JsonConstructor]
        public RALColorListNodeModel(IEnumerable<PortModel> inPorts, IEnumerable<PortModel> outPorts) : base(inPorts, outPorts)
        {
            this.PortDisconnected += RALList_PortDisconnected;
        }

        // Restore default button/window text and trigger UI update
        private void RALList_PortDisconnected(PortModel obj)
        {
            RALColorList = defaultRALColorList;
            RaisePropertyChanged("RALColorList");
        }

        #endregion

        //Use the VMDataBridge to safely retrieve our input values
        #region Data bridge Call back
        /// <summary>
        /// Register the data bridge callback.
        /// </summary>
        protected override void OnBuilt()
        {
            base.OnBuilt();
            VMDataBridge.DataBridge.Instance.RegisterCallback(GUID.ToString(), DataBridgeCallback);
        }

        /// <summary>
        /// Unregister the data bridge callback.
        /// </summary>
        public override void Dispose()
        {
            base.Dispose();
            VMDataBridge.DataBridge.Instance.UnregisterCallback(GUID.ToString());
        }

        /// <summary>
        /// Callback method for DataBridge mechanism.
        /// This callback only gets called when 
        ///     - The AST is executed
        ///     - After the BuildOutputAST function is executed 
        ///     - The AST is fully built
        /// </summary>
        /// <param name="data">The data passed through the data bridge.</param>
        private void DataBridgeCallback(object data)
        {
            ArrayList inputs = data as ArrayList;
            var inputList = new ObservableCollection<RALColor>();

            foreach (var input in inputs)
            {
                if (input.GetType() == typeof(string))
                    inputList.Add(RALColorHelpers.GetRALColorByName(input.ToString()));

                else
                    foreach (var i in (ArrayList)input)
                        inputList.Add(RALColorHelpers.GetRALColorByName(i.ToString()));
            }

            RALColorList = inputList;
        }
        #endregion

        #region public methods
        /// <summary>
        /// BuildOutputAst is where the outputs of this node are calculated.
        /// This method is used to do the work that a compiler usually does 
        /// by parsing the inputs List inputAstNodes into an abstract syntax tree.
        /// </summary>
        /// <param name="inputAstNodes"></param>
        /// <returns></returns>
        [IsVisibleInDynamoLibrary(false)]
        public override IEnumerable<AssociativeNode> BuildOutputAst(List<AssociativeNode> inputAstNodes)
        {
            if (!InPorts[0].IsConnected)
            {
                return new[]
                {
                    AstFactory.BuildAssignment(GetAstIdentifierForOutputIndex(0), AstFactory.BuildNullNode()),
                };
            }

            var args = new List<AssociativeNode> { inputAstNodes[0] };
            var lboxFuncNode = AstFactory.BuildFunctionCall(
                new Func<string, Color>(RALColorHelpers.GetDynColorFromRAL), args);
           
            return new[]
            {
                AstFactory.BuildAssignment(GetAstIdentifierForOutputIndex(0), lboxFuncNode)
                    ,AstFactory.BuildAssignment(
                        AstFactory.BuildIdentifier(AstIdentifierBase + "_dummy"),
                        VMDataBridge.DataBridge.GenerateBridgeDataAst(GUID.ToString(), AstFactory.BuildExprList(inputAstNodes)))
            };
        }
        #endregion
    }
}
```
```
using System.Collections.Generic;
using System.Windows.Controls;
using Dynamo.Controls;
using Dynamo.Models;
using Dynamo.ViewModels;
using Dynamo.Wpf;
using Dynamo.Configuration;

namespace RaDy.CNM.ListBoxNodes
{
    /// <summary>
    /// View customizer for CustomNodeModel Node Model.
    /// </summary>
    public class RALColorListNodeViewCustomization : INodeViewCustomization<RALColorListNodeModel>
    {
        /// <summary></summary>
        /// <param name="model">The NodeModel representing the node's core logic.</param>
        /// <param name="nodeView">The NodeView representing the node in the graph.</param>
        public void CustomizeView(RALColorListNodeModel model, NodeView nodeView)
        {
            var listBox = new RALColorListBox();
            
            nodeView.inputGrid.Children.Add(listBox);
            listBox.DataContext = model;

        }

        /// <summary>
        /// Here you can do any cleanup you require if you've assigned callbacks for particular 
        /// UI events on your node.
        /// </summary>
        public void Dispose() { }
    }
}
```
@mjkkirschner @ksobon 

Could you please help/explain why this is happening and what should i do in order to see this node even after saving. Thanks in advance.

**Comment by mjkkirschner** [comment]
read this first if you have not already:
https://developer.dynamobim.org/03-Development-Options/upgrade-nodes-for-dynamo-20.html

The nodelModel is not being deserialized correctly - when you look at your saved json what do you see for your node?

I'm guessing there might some issue automatically deserializing your public observableCollection because it's of a complex type. You'll probably want to mark that jsonignore and use a backing collection of simple color objects or strings or something.

Have you you attached the debugger to see what exception gets raised when you try opening your saved file?

**Comment by argram30** [comment]
@mjkkirschner many thanks for your prompt reply. Yes I did read the link you mentioned. 

As per your suggestion I first tried including the attribute [JsonIgnore] for observableCollection property that, indeed did solve the problem and node was able to deserialize and appear properly on the graph. Thanks once again for pointing me to the right mistake. 

Next time I will try opening the dynamo file to see how its serialized. Have a good day.
